google/cloud/storage/status_or.h

```diff
@@ -0,0 +1,178 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_STATUS_OR_H_
+#define GOOGLE_CLOUD_CPP_GOOGLE_CLOUD_STORAGE_STATUS_OR_H_
+
+#include "google/cloud/internal/throw_delegate.h"
+#include "google/cloud/storage/internal/throw_status_delegate.h"
+#include "google/cloud/storage/status.h"
+#include <type_traits>
+#include <utility>
+
+namespace google {
+namespace cloud {
+namespace storage {
```

Comment:
I imagine we'll want this (or a similar) class for all our future APIs. So do you think it's worth start off with this thing in the google3/cloud/ directory so it's not GCS-specific?

Reply:
Eventually, yes.

To do this now we need to move `storage::Status` too, and I am not certain we want the class as-is outside `google/cloud/storage/` because:

* `storage::Status` right now returns the http error code (404, 200, etc.) we do not expose those to the application (the exceptions hide them), but I would rather use the typical `StatusCode` stuff before we expose these in a "generic" class.
* all the other services use gRPC, and there is an existing `grpc::Status`.
  * Would we map `grpc::Status` to `google::cloud::Status`?  Consider that there is no ref-ref accessors for `grpc::Status`, so it would be copies everywhere.
  * If not, then what do we do with GCS that does not have `grpc::Status`?  Do we link the gRPC library just for that class?
* Moving `google::cloud::storage::Status` to `google::cloud::Status` requires a change to all the functions in `google/cloud/storage/internal/`, maybe not the right thing to do right now.